proc-macros: rename parameter names

deflator-hacker opened the pull request:
Decoupling rust parameter names from (de)serialized json is essential for me and useful in general.

Some usecases include:

- supporting parameter names that are reserved keywords in rust, such as `type` (currently impossible)
- talking to case sensitive RPC servers only accepting non-snake-case parameters (currently impossible without clippy complaining)

Those are both usecases I personally have.

The PR implements this by introducing a new attribute for the `rpc` macro called `argument`.

- documentation has been updated
- tests have been written

I previously responded to an outdated similar PR [here](https://github.com/paritytech/jsonrpsee/pull/1015#issuecomment-2094789210).

Minimal working example:

```rust
#[rpc(client, server)]
pub trait MyRpc {
	#[method(name = "foo")]
	async fn foo(
		&self,
		param_a: u8,
		#[argument(rename = "type")]
		r#type: String
	) -> RpcResult<u16>;
}
```

I divided the changes into multiple commits because they are a bit lengthy.
This is my first open source contribution, I am eager to learn, let me know what can be improved.

deflator-hacker committed:
proc-macros: add RpcFnArg to parse function argument attribute

deflator-hacker committed:
proc-macros: integate RpcFnArg into rpc macro

deflator-hacker committed:
proc-macros: update rpc macro documentation example to display argument attribute

deflator-hacker committed:
proc-macros: add tests to verify correct argument rename logic

deflator-hacker committed:
proc-macros: add more documentation on rpc macro for argument attribute

deflator-hacker committed:
proc-macros: fix tabs in doc comments

deflator-hacker committed:
proc-macros: fix lints generated by CI

niklasad1 commented:
Looks overall good, nice job

I'm not following why you need to rename parameters when those passed as Array (these are omitted by the client)? In such scenarios then variable doesn't correspond to the actual param name, so why is it needed?

The contrary applies to JSON-Object where the key is the same name passed in the call with some glue to convert snake_case to camaleCase. Then I follow why these need to be renamed to support reserved rust keywords etc....

I'll take a closer look tomorrow.

deflator-hacker commented:
> I'm not following why you need to rename parameters when those passed as Array (these are omitted by the client)? In such scenarios then variable doesn't correspond to the actual param name, so why is it needed?

You are absolutely right, there is no point (actually no way) to rename parameters in arrays.
I included a check for array params in [proc_macros/tests/ui/correct/rename_fn_params.rs](https://github.com/deflator-hacker/jsonrpsee/blob/master/proc-macros/tests/ui/correct/rename_fn_params.rs#L17) just because I wanted to assert that I did not break anything deserialization-related by accident.

niklasad1 commented on proc-macros/src/render_server.rs:
```diff
@@ -147,25 +150,23 @@ impl RpcDescription {
-								#parsing
-								#into_response::into_response(context.as_ref().#rust_method_name(#params_seq).await)
-							})
+				} else if method.signature.sig.asyncness.is_some() {
```
I suppose this is unrelated and just fixes a clippy warning?

deflator-hacker replied:
Yes, I can revert if this is not the right place to do this but CI tests linted this so I fixed it.

niklasad1 commented on proc-macros/src/rpc_macro.rs:
```diff
@@ -37,13 +37,57 @@ use quote::quote;
 use syn::spanned::Spanned;
 use syn::{punctuated::Punctuated, Attribute, Token};
 
+/// argument to rpc function
```
```suggestion
/// Represents a single argument in a RPC call.
```

deflator-hacker replied:
done

niklasad1 commented on proc-macros/src/rpc_macro.rs:
```diff
@@ -37,13 +37,57 @@ use quote::quote;
+/// stores modifications based on attributes
+#[derive(Debug, Clone)]
+pub struct RpcFnArg {
+	pub arg_pat: syn::PatIdent,
```
can you make these private because you have added getters fns below?

niklasad1 replied:
nvm, it's used to extract the ty in a couple of places

niklasad1 commented on proc-macros/tests/ui/correct/basic.rs:
```diff
@@ -111,6 +122,8 @@ async fn main() {
 
 	assert_eq!(client.async_method(10, "a".into()).await.unwrap(), 42);
+	assert_eq!(client.rename_params(256, true).await.unwrap(), 128);
+	assert_eq!(client.rename_params(256, false).await.unwrap(), 256);
```
you can just add two additional tests here with object params.

```rust
	let mut params = ObjectParams::new();
	params.insert("type", 256).unwrap();
	params.insert("halfType", false).unwrap();

	assert_eq(client.request::<u16, ObjectParams>("rename_params", params).await.unwrap(), 128);
```

```rust
	let mut params = ObjectParams::new();
	params.insert("type", 256).unwrap();
	params.insert("halfType", true).unwrap();

	assert_eq(client.request::<u16, ObjectParams>("rename_params", params).await.unwrap(), 256);
```

Then remove the extra test file that you added below "proc-macros/tests/ui/correct/rename_fn_params.rs"

deflator-hacker replied:
done

niklasad1 commented on proc-macros/src/rpc_macro.rs:
```diff
@@ -37,13 +37,57 @@ use quote::quote;
+		&self.arg_pat
+	}
+	/// Return the string representation of this argument when (de)serializing.
+	pub fn rename_to(&self) -> &Option<String> {
```
is it possible to refactor this method then internally fetch the name from the rename if it exist otherwise fall back to self.arg_pat?

It would be easier to just call `fn_arg.name()`

```rust
pub fn name(&self) -> String {
	self.rename_to.clone().unwrap_or_else(|| self.arg_pat.ident.to_string())
}
```

deflator-hacker replied:
done

niklasad1 commented on proc-macros/src/render_server.rs:
```diff
@@ -383,18 +384,30 @@ impl RpcDescription {
+					)
 				));
 
+				let serde_rename = rename_to
```
we shouldn't need this serde rename would be better to rename the name in the struct instead, if you agree with my proposed API with a function name we could do this instead:

```rust
			let fields = params.iter().zip(generics.clone()).map(|(fn_arg, ty)| {
				// this could be a renamed name or not... but shouldn't matter at this point
                let name = fn_arg.name();

				let mut alias_vals = String::new();
				alias_vals.push_str(&format!(r#"alias = "{}""#, heck::ToSnakeCase::to_snake_case(name.as_str())));
				alias_vals.push(',');
				alias_vals
					.push_str(&format!(r#"alias = "{}""#, heck::ToLowerCamelCase::to_lower_camel_case(name.as_str())));

				let alias = TokenStream2::from_str(alias_vals.as_str()).unwrap();
				let serde_alias: Attribute = syn::parse_quote! {
					#[serde(#alias)]
				};
				let unquoted_name = TokenStream2::from_str(name.as_str()).unwrap();

				quote! {
					#serde_alias
					#unquoted_name: #ty,
				}
			});
```

deflator-hacker replied:
I adopted the `RpcFnArg::name()` API.

I tested this change, it does not work because `RpcFnArg::name()` can be _any_ string, such as `"type"`, `"in"`, `"for"` etc.
If it was `"type"` the struct would be invalid because `"type"` is a reserved keyword.
Leveraging serde here gives us all possible strings that can be used with serde.

deflator-hacker committed:
proc-macros: update internal documentation on RpcFnArg

Co-authored-by: Niklas Adolfsson <niklasadolfsson1@gmail.com>

deflator-hacker committed:
proc-macros: use ObjectParams to test deserialization

deflator-hacker committed:
proc-macros: remove rename_fn_params test

deflator-hacker committed:
proc-macros: fix basic tests by importing ObjectParams and respecting namespace

deflator-hacker committed:
proc-macros: refactor RpcFnArg

- expose potentially renamed name as String insead of Option in RpcFnArg::name()
- make RpcFnArg::rename_to private to migrate to new RpcFnArg::name()
- adapt render_client.rs and render_server.rs to changes

niklasad1 commented:
Looks great, I'll can push a fix to unblock the UI tests tomorrow it's probably the latest rustc release that broke them.

niklasad1 commented on proc-macros/tests/ui/correct/basic.rs:
```diff
@@ -24,6 +24,13 @@ pub trait Rpc {
 	#[method(name = "array_params")]
 	async fn array_params(&self, items: Vec<u64>) -> RpcResult<u64>;
 
+	#[method(name = "rename_params", param_kind = map)]
```
you shouldn't need "param_kind = map" because it's only used when you are using the proc macro API anyway it doesn't matter.

deflator-hacker replied:
thanks, removed it

niklasad1 commented on proc-macros/tests/ui/correct/basic.rs:
```diff
@@ -111,6 +122,17 @@ async fn main() {
+	let mut params = ObjectParams::new();
+	params.insert("type", 256).unwrap();
+	params.insert("halfType", true).unwrap();
+	assert_eq!(client.request::<u16, ObjectParams>("foo_rename_params", params).await.unwrap(), 128);
```
can you add another call here with ArrayParams to ensure that the rename stuff doesn't break it?

deflator-hacker replied:
added

deflator-hacker committed:
proc-macros: update basic ui test

- remove unnecessary param_kind: map
- add array params test to prevent accidental breakage through param renaming

niklasad1 commented on proc-macros/src/render_server.rs:
```diff
@@ -383,17 +384,17 @@ impl RpcDescription {
+				alias_vals
+					.push_str(&format!(r#"alias = "{}""#, heck::ToLowerCamelCase::to_lower_camel_case(name.as_str())));
+
+				let serde_rename = quote!(#[serde(rename = #name)]);
```
This is needed because arg_pat below can't be reserved rust names such as type

niklasad1 commented on proc-macros/src/rpc_macro.rs:
```diff
@@ -37,13 +37,57 @@ use quote::quote;
+/// stores modifications based on attributes
+#[derive(Debug, Clone)]
+pub struct RpcFnArg {
+	pub arg_pat: syn::PatIdent,
```
```suggestion
	pub(crate) arg_pat: syn::PatIdent,
```

niklasad1 committed:
Update proc-macros/src/rpc_macro.rs

niklasad1 commented on proc-macros/src/rpc_macro.rs:
```diff
@@ -37,13 +37,57 @@ use quote::quote;
+pub struct RpcFnArg {
+	pub(crate) arg_pat: syn::PatIdent,
+	rename_to: Option<String>,
+	pub ty: syn::Type,
```
```suggestion
	pub(crate) ty: syn::Type,
```

niklasad1 committed:
Update proc-macros/src/rpc_macro.rs

niklasad1 reviewed (approved):
Nice work looks good

deflator-hacker committed:
Merge branch 'paritytech:master' into master

jsdw commented on proc-macros/src/lib.rs:
```diff
@@ -206,6 +206,16 @@ pub(crate) mod visitor;
+///
+/// - `rename`: rename the generated JSON key.
+///
+/// **Method requirements:**
```
This line isn't followed by anything? :)

```suggestion
```

jsdw commented on proc-macros/tests/ui/correct/basic.rs:
```diff
@@ -24,6 +24,13 @@ pub trait Rpc {
+	#[method(name = "rename_params")]
+	async fn rename_params(
+		&self,
+		#[argument(rename = "type")] r#type: u16,
```
DQ: without this rename, would it not still be valid to pass a param with name "type"?

deflator-hacker replied:
I am unsure what you are asking.
I think that it would not be valid because the server would expect params with name "r#type" instead of "type".

Actually, the original usecase that got me to writing this PR was that I had a non-jsonrpsee server accepting a param "in" which could not be done through the proc-macros API because "in" is a reserved keyword (as "type" is, too) and thus all params would be serialized to "r#in", preventing the server from recognizing the param.

niklasad1 replied:
So the rpc definition looks like this:

```rust
    #[method(name = "foo")]
    fn bar(f: usize, r#type: usize) -> usize {
        f + r#type
    }
```

Then in codegen we generated a struct as JSON Object with param names as the idents from the params binding such that the following would be generated:

```rust
    #[derive(serde::Serialize, serde::Deserialize)
    struct ParamsObject<G0, G1> {
        #[serde(alias = "storage_key", alias = "storageKey")]
        f: G0,
        #[serde(alias = "r_type", alias = "rType")]
        r#type: G1,
    }
```

Thus, it was not possible to make call to the server with  `type` as parameter name because it's keyword in rust prior to this PR. All good

jsdw replied:
Aah thanks for the explanations! I thought that `r#type` might just resolve to the ident `type` anyway (ignoring the raw prefix bit that allows you to actually type it at all), but I was wrong actually about that :)

jsdw commented on proc-macros/tests/ui/correct/basic.rs:
```diff
@@ -24,6 +24,13 @@ pub trait Rpc {
+	async fn rename_params(
+		&self,
+		#[argument(rename = "type")] r#type: u16,
+		#[argument(rename = "halfType")] half_type: bool,
```
I don't recall here either; would `halfType` not already be accepted or do we not to any camel cas transform?

I'd feel more confident that the test was working if the rename was to something more different than the original :)

deflator-hacker replied:
You are right for servers generated by jsonrpsee, they accept mutliple cases ([see here](https://github.com/paritytech/jsonrpsee/blob/master/proc-macros/src/render_server.rs#L390)).

When talking to other JSON-RPC servers this might not be the case, hence my PR.

jsdw reviewed (approved):
LGTM, just a couple of nits!

deflator-hacker committed:
proc-macros: remove unnecessary docs

Co-authored-by: James Wilson <james@jsdw.me>

niklasad1 committed:
minor clarification in tests
